### PR TITLE
Remove `QualifiedDeclNameSyntax`

### LINKED ISSUE #1975: Remove `QualifiedDeclNameSyntax` in favor of `MemberAccessExprSyntax`
`QualifiedDeclNameSyntax` is only used for the `@derivative(of:)` attribute. Everywhere else we model it as a `MemberAccessExprSyntax`. We should do the same for `@deriviate(of:)` and remove the `QualifiedDeclNameSyntax` node.

### PR DESCRIPTION
**Pull request by kimdv**
Resolves #1975

I have been looking a bit into this and there is some differences in the models of `QualifiedDeclNameSyntax` and `MemberAccessExprSyntax`.

Etc the `period` is optional in `QualifiedDeclNameSyntax` and required in `MemberAccessExprSyntax`. 
This makes some tests failing. 

```swift
 @transpose(of: +)
      func addTranspose(_ v: Float) -> (Float, Float) {
        return (v, v)
      }
```

or 

```swift
@inlinable
      @derivative(of: differentiableReduce)
      internal func _vjpDifferentiableReduce<Result: Differentiable>(
        _ initialResult: Result,
        _ nextPartialResult: @differentiable(reverse) (Result, Element) -> Result
      ) -> (
        value: Result,
        pullback: (Result.TangentVector)
          -> (Array.TangentVector, Result.TangentVector)
      ) {}
```

### CODE REVIEW

**Review by ahoppen — APPROVED** [review]
LGTM. I am not entirely sure if the current implementation of `parseQualifiedDeclName` is correct but this certainly doesn’t make it less correct. So: Good to go

**Comment by kimdv** [comment]
> LGTM. I am not entirely sure if the current implementation of `parseQualifiedDeclName` is correct but this certainly doesn’t make it less correct. So: Good to go

I think it is not correct. 
As mentioned then the is some difference on `QualifiedDeclNameSyntax` model and `MemberAccessExprSyntax` model. 

Currently the tests are failing.

The problem is that in `MemberAccessExprSyntax` is required. Previously is was an optional type

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 1]
```diff
@@ -1205,7 +1206,8 @@ public let EXPR_NODES: [Node] = [
       Child(
         name: "DeclName",
         kind: .node(kind: .declReferenceExpr),
-        nameForDiagnostics: "name"
+        nameForDiagnostics: "name",
+        documentation: "The name of the referenced function."
```
[the remark this answers is not in the document]
Comment (kimdv) [reply CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift]: Changed the phrase a bit

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 2]
```diff
@@ -1195,6 +1195,7 @@ public let EXPR_NODES: [Node] = [
         name: "Base",
         kind: .node(kind: .expr),
         nameForDiagnostics: "base",
+        documentation: "The base type of the qualified name, optionally specified.",
```
[the remark this answers is not in the document]
Comment (kimdv) [reply CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift]: Changed the phrase a bit

**Comment by ahoppen** [comment]
> I'm not sure but I think that we should change to ExprSyntax so we also can use IdentifierSyntax etc when we have a cases like @derivative(of: differentiableReduce) or @transpose(of: +)

I think that would be a good solution.

**Comment by kimdv** [comment]
@ahoppen rephrased the documentation and changed to ExprSyntax. 

Can you re-review?

**Comment by kimdv** [comment]
@swift-ci please test

**Review by ahoppen — APPROVED** [review]

**Pull request merged**